### PR TITLE
Hold cs_main when reading chainActive via getTipLocator(). Remove assumeLocked().

### LINKED ISSUE #16028: getTipLocator() reads chainActive without holding cs_main despite LockAnnotation lock(::cs_main) guarantee
`LockAnnotation lock(::cs_main)` is a guarantee to the compiler thread analysis that `::cs_main` is locked (when it couldn't be determined otherwise).

Despite being annotated with the locking guarantee ...

https://github.com/bitcoin/bitcoin/blob/65526fc8666fef35ef908dbc225f706bef642c7e/src/interfaces/chain.cpp#L134-L138

... `getTipLocator()` reads `chainActive` (via `::ChainActive()`) without holding `cs_main`.

This can be verified by adding the following `AssertLockHeld(cs_main)`:

```
$ git diff
diff --git a/src/interfaces/chain.cpp b/src/interfaces/chain.cpp
index 59623284d..9fc693a0f 100644
--- a/src/interfaces/chain.cpp
+++ b/src/interfaces/chain.cpp
@@ -134,6 +134,7 @@ class LockImpl : public Chain::Lock
     CBlockLocator getTipLocator() override
     {
         LockAnnotation lock(::cs_main);
+        AssertLockHeld(::cs_main);
         return ::ChainActive().GetLocator();
     }
     Optional<int> findLocatorFork(const CBlockLocator& locator) override
$ make check
../build-aux/test-driver: line 107: 12881 Aborted                 "$@" > $log_file 2>&1
FAIL: qt/test/test_bitcoin-qt
```

### PR DESCRIPTION
**Pull request by practicalswift**
Fixes #16028.

Problem description:

`LockAnnotation lock(::cs_main)` is a guarantee to the compiler thread analysis that `::cs_main` is locked (when it couldn't be determined otherwise).

Despite being annotated with the locking guarantee ...

https://github.com/bitcoin/bitcoin/blob/65526fc8666fef35ef908dbc225f706bef642c7e/src/interfaces/chain.cpp#L134-L138

... `getTipLocator()` reads `chainActive` (via `::ChainActive()`) without holding `cs_main`.

This can be verified by adding the following `AssertLockHeld(cs_main)`:

```
$ git diff
diff --git a/src/interfaces/chain.cpp b/src/interfaces/chain.cpp
index 59623284d..9fc693a0f 100644
--- a/src/interfaces/chain.cpp
+++ b/src/interfaces/chain.cpp
@@ -134,6 +134,7 @@ class LockImpl : public Chain::Lock
     CBlockLocator getTipLocator() override
     {
         LockAnnotation lock(::cs_main);
+        AssertLockHeld(::cs_main);
         return ::ChainActive().GetLocator();
     }
     Optional<int> findLocatorFork(const CBlockLocator& locator) override
$ make check
../build-aux/test-driver: line 107: 12881 Aborted                 "$@" > $log_file 2>&1
FAIL: qt/test/test_bitcoin-qt
```

### CODE REVIEW

**Comment by ryanofsky** [comment]
Can you also delete the LockingStateImpl class? There is no reason for the LockingStateImpl class to exist without assumeLocked. You just need to s/LockingStateImpl/LockImpl/ in the lock method and make LockImpl inherit from UniqueLock.

**Comment by practicalswift** [comment]
@ryanofsky Thanks for the quick review!

Do you mean like this?

```diff
diff --git a/src/interfaces/chain.cpp b/src/interfaces/chain.cpp
index 1a8c32ee3..eaddcd9dd 100644
--- a/src/interfaces/chain.cpp
+++ b/src/interfaces/chain.cpp
@@ -37,7 +37,7 @@
 namespace interfaces {
 namespace {

-class LockImpl : public Chain::Lock
+class LockImpl : public Chain::Lock, public UniqueLock<CCriticalSection>
 {
     Optional<int> getHeight() override
     {
@@ -155,10 +155,7 @@ class LockImpl : public Chain::Lock
         return AcceptToMemoryPool(::mempool, state, tx, nullptr /* missing inputs */, nullptr /* txn replaced */,
             false /* bypass limits */, absurd_fee);
     }
-};

-class LockingStateImpl : public LockImpl, public UniqueLock<CCriticalSection>
-{
     using UniqueLock::UniqueLock;
 };

@@ -249,10 +246,10 @@ class ChainImpl : public Chain
 public:
     std::unique_ptr<Chain::Lock> lock(bool try_lock) override
     {
-        auto result = MakeUnique<LockingStateImpl>(::cs_main, "cs_main", __FILE__, __LINE__, try_lock);
+        auto result = MakeUnique<LockImpl>(::cs_main, "cs_main", __FILE__, __LINE__, try_lock);
         if (try_lock && result && !*result) return {};
         // std::move necessary on some compilers due to conversion from
-        // LockingStateImpl to Lock pointer
+        // LockImpl to Lock pointer
         return std::move(result);
     }
     bool findBlock(const uint256& hash, CBlock* block, int64_t* time, int64_t* time_max) override
```

**Comment by ryanofsky** [comment]
> Do you mean like this?

Thanks, yes that looks perfect.

File: src/wallet/test/wallet_tests.cpp [hunk 1]
```diff
@@ -399,8 +401,9 @@ BOOST_FIXTURE_TEST_CASE(ListCoins, ListCoinsTestingSetup)
     // address.
     std::map<CTxDestination, std::vector<COutput>> list;
     {
+        auto locked_chain = m_chain->lock();
         LOCK2(cs_main, wallet->cs_wallet);
```
Comment (MarcoFalke) [inline src/wallet/test/wallet_tests.cpp]:
```suggestion
        LOCK(wallet->cs_wallet);
```

**Comment by MarcoFalke** [comment]
Could squash all the refactoring except the change to /wallet/wallet.cpp, which could go into a separate commit?

**Comment by practicalswift** [comment]
@MarcoFalke Thanks for the quick review! Updated accordingly. Please re-review :-)

**Comment by MarcoFalke** [comment]
utACK 9402ef0739fdcd8e989c07c0595095e9608b243c

**Review by ryanofsky — APPROVED** [review]
utACK 00f941055fedb855473d5428f54c9d79786f6e23

**Comment by practicalswift** [comment]
@ryanofsky Thanks for the review. I think you utACK:ed the previous commit hash :-)

**Review by ryanofsky — APPROVED** [review]
utACK 9402ef0739fdcd8e989c07c0595095e9608b243c. Changes are consolidating commits and removing redundant lock2 cs_main calls

**Pull request merged**